### PR TITLE
feat: verify provenance

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>
Co-authored-by: Furkan Türkal <furkan.turkal@trendyol.com>

PTAL @laurentsimon @ianlewis

### CODE REVIEW

File: .github/workflows/goreleaser.yml [hunk 1]
```diff
@@ -24,10 +24,10 @@ jobs:
         uses: actions/checkout@24cb9080177205b6e8c946b17badbe402adc938f # v3.4.0
         with:
           fetch-depth: 0
-      - name: Run Tests
-        uses: ./.github/actions/test.yaml
-      - name: Run Lints
-        uses: ./.github/actions/lint.yaml
+      # - name: Run Tests
```
Comment (oliverchang) [inline .github/workflows/goreleaser.yml:27]: Can we bring these back? This is an important part of the release process.

File: .github/workflows/goreleaser.yml [hunk 2]
```diff
@@ -67,3 +71,43 @@ jobs:
...
+    steps:
+      - name: Install SLSA verifier
+        uses: slsa-framework/slsa-verifier/actions/installer@v2.1.0
+      - name: Download assets
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PROVENANCE: "${{ needs.provenance.outputs.provenance-name }}"
+        run: |
+          set -euo pipefail
+          TAG=${GITHUB_REF#refs/tags/}
```
Comment (oliverchang) [inline .github/workflows/goreleaser.yml:88]: There's a fair bit of bash here to maintain :) 

@laurentsimon @ianlewis is this recommended practice? If so, would it be possible at all to generalize this in a more re-usable way "upstream" somewhere as part of the slsa-framework repo?

**Review by oliverchang — COMMENTED** [review]
thanks for this contribution! just some questions for both you and @laurentsimon and @ianlewis :)

File: .github/workflows/goreleaser.yml (hunk 2 above)
Comment (ianlewis) [reply .github/workflows/goreleaser.yml:88]: Yeah, it might be a good idea to have an action for verifying artifacts on a GitHub release. i.e. Given a release download all the artifact and verify them.

This would only work for a situation like osv-scanner where you provide binary artifacts along with the provenance (as opposed to some other language package) but I think it could still be useful.

We would also need to be careful to provide a way to provide the list of expected artifacts. We would need to catch the situation where an artifact was excluded from the provenance even though it was expected to be there (in this case `goreleaser.outputs.hashes`)

/cc @laurentsimon

File: .github/workflows/goreleaser.yml [hunk 3]
```diff
@@ -49,12 +49,16 @@ jobs:
...
+          ACTOR: ${{ github.actor }}
       - name: Generate subject
         id: hash
         env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           ARTIFACTS: "${{ steps.run-goreleaser.outputs.artifacts }}"
         run: |
           set -euo pipefail
+          gh api "repos/developer-guy/osv-scanner/releases"
+          echo $GITHUB_REPOSITORY
```
Comment (ianlewis) [inline .github/workflows/goreleaser.yml:61]: I assume this is debugging code and you'll want to remove it?

File: .goreleaser.yml [hunk 4]
```diff
@@ -32,7 +32,7 @@ builds:
 dockers:
   # Arch: amd64
   - image_templates:
-      - 'ghcr.io/google/osv-scanner:{{ .Tag }}-amd64'
+      - 'ghcr.io/{{ .Env.ACTOR }}/osv-scanner:{{ .Tag }}-amd64'
```
Comment (ianlewis) [inline .goreleaser.yml:35]: Is this right? I'm not sure that the `github.actor` is going to be right if the repo is owned by an organization.

File: .github/workflows/goreleaser.yml [hunk 5]
```diff
@@ -67,3 +71,43 @@ jobs:
...
+          PROVENANCE: "${{ needs.provenance.outputs.provenance-name }}"
+        run: |
+          set -euo pipefail
+          TAG=${GITHUB_REF#refs/tags/}
+          echo "Downloading assets for tag $TAG"
+          for i in $(gh api "repos/$GITHUB_REPOSITORY/releases" | jq -r '.[] | select(.tag_name=='\"$TAG\"').assets[].id')
+          do
+            echo "Downloading asset $i"
+            curl -L -o $(gh api "repos/$GITHUB_REPOSITORY/releases/assets/$i" | jq -r '.name') -H "Authorization: token $GH_TOKEN" -H "Accept: application/octet-stream" $(gh api "repos/$GITHUB_REPOSITORY/releases/assets/$i" | jq -r '.url')
+          done
```
Comment (ianlewis) [inline .github/workflows/goreleaser.yml:94]: Is there a reason you can't just do `gh release download "$TAG"` ?

File: .github/workflows/goreleaser.yml [hunk 6]
```diff
@@ -67,3 +71,43 @@ jobs:
...
+          echo "Downloading provenance $PROVENANCE"
+          gh -R $GITHUB_REPOSITORY release download --clobber $GITHUB_REF_NAME -p $PROVENANCE
```
[the remark this answers is not in the document]
Comment (laurentsimon) [reply .github/workflows/goreleaser.yml:96]: I think the code from https://github.com/google/go-containerregistry/blob/main/.github/workflows/release.yml#L48-L76 should work just fine, no?

File: .github/workflows/goreleaser.yml (hunk 2 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml:88]: I can work on this 🤩

File: .github/workflows/goreleaser.yml (hunk 1 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml:27]: the reason why I did this is that workflow didn't worked as it didn't find these action files under `.github/actions`.

File: .goreleaser.yml (hunk 4 above)
Comment (developer-guy) [reply .goreleaser.yml:35]: I might use `github.repository_owner` instead, will it work?

File: .github/workflows/goreleaser.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (developer-guy) [reply .github/workflows/goreleaser.yml:96]: here is the PR for that: _https://github.com/google/go-containerregistry/pull/1611_

`gh release download` didn't work with draft releases as osv-scanner marks its release as draft and I looked for a solution for `gh release download` to enable downloading assets from draft release but I couldn't find anything useful, then I turned the solution back to the GitHub API calls via `gh api`. But let me try again to ensure that I didn't work.

File: .github/workflows/goreleaser.yml (hunk 5 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml:94]: for the same reason I explained above: _https://github.com/google/osv-scanner/pull/309#discussion_r1146679828_

File: .github/workflows/goreleaser.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (developer-guy) [reply .github/workflows/goreleaser.yml]: it worked btw with the right permissions, so to download releases it requires you to put write permissions for content. Great!

File: .github/workflows/goreleaser.yml (hunk 5 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml]: it worked @ianlewis with right permisssions.

File: .github/workflows/goreleaser.yml (hunk 1 above)
Comment (another-rex) [reply .github/workflows/goreleaser.yml:27]: Ah my bad, fixing this in #311.

File: .goreleaser.yml (hunk 4 above)
Comment (developer-guy) [reply .goreleaser.yml:35]: I renamed it as `OWNER` and used `github.repository_owner`.

File: .github/workflows/goreleaser.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (laurentsimon) [reply .github/workflows/goreleaser.yml]: that's interesting. Maybe file an issue on GitHub.. sounds like a bug... they should have a REST flag to allow for draft release to be fetched, without the need for a write token.

File: .github/workflows/goreleaser.yml [hunk 7]
```diff
@@ -67,3 +69,36 @@ jobs:
     with:
       base64-subjects: "${{ needs.goreleaser.outputs.hashes }}"
       upload-assets: true # upload to a new release
+  verification:
+    needs: [goreleaser, provenance]
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write # To add assets to a release.
```
Comment (laurentsimon) [inline .github/workflows/goreleaser.yml:76]: please file a bug on https://github.com/actions/runner and cc me. Requiring write permission is really odd

**Review by laurentsimon — APPROVED** [review]
LGTM

File: .github/workflows/goreleaser.yml (hunk 7 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml]: let me try it again, I changed back it to "read".

File: .github/workflows/goreleaser.yml (hunk 1 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml]: @another-rex I think the problem still continues. I also do not think that this is the way of referencing different workflows within the workflow.

Screen Shot 2023-03-25 at 3 29 07 PM

File: .github/workflows/goreleaser.yml (hunk 7 above)
Comment (developer-guy) [reply .github/workflows/goreleaser.yml]: see it didn't worked: _https://github.com/developer-guy/osv-scanner/actions/runs/4519224365/jobs/7959640924_

Comment (developer-guy) [reply .github/workflows/goreleaser.yml:76]: it didn't work with "write" permission either: 
_https://github.com/developer-guy/osv-scanner/actions/runs/4519341906/jobs/7959797410#step:1:17_

but it worked once what is going on 🤔 
_https://github.com/developer-guy/osv-scanner/actions/runs/4504207210/jobs/7928457868#step:1:17_

Comment (developer-guy) [reply .github/workflows/goreleaser.yml:76]: this script is just working fine with `write` permission: 

```bash
set -euo pipefail
  TAG=${GITHUB_REF#refs/tags/}
  echo "Downloading assets for tag $TAG"
  for i in $(gh api "repos/$GITHUB_REPOSITORY/releases" | jq -r '.[] | select(.tag_name=='\"$TAG\"').assets[].id')
  do
    echo "Downloading asset $i"
    curl -L -o $(gh api "repos/$GITHUB_REPOSITORY/releases/assets/$i" | jq -r '.name') -H "Authorization: token $GH_TOKEN" -H "Accept: application/octet-stream" $(gh api "repos/$GITHUB_REPOSITORY/releases/assets/$i" | jq -r '.url')
  done
  echo "Downloading provenance $PROVENANCE"
  gh -R $GITHUB_REPOSITORY release download --clobber $GITHUB_REF_NAME -p $PROVENANCE
```

Comment (developer-guy) [reply .github/workflows/goreleaser.yml:76]: we even don't need this line:

```bash
gh -R $GITHUB_REPOSITORY release download --clobber $GITHUB_REF_NAME -p $PROVENANCE
```

and I also disabled uploading assets in the generic generator settings:

```bash
upload-assets: false
```

File: .github/workflows/goreleaser.yml (hunk 1 above)
Comment (another-rex) [reply .github/workflows/goreleaser.yml:27]: You're right, I misunderstood how the uses line works. I updated and tested it in #315, which seems to work in my fork.

File: .github/workflows/goreleaser.yml (hunk 2 above)
Comment (oliverchang) [reply .github/workflows/goreleaser.yml:86]: @ianlewis @laurentsimon thoughts on this going into some slsa-framework repo instead (either a new one or existing?)

 I would very much prefer that approach as it makes this work a lot more useful and we (OSV-Scanner maintainers) have less things we're not as familiar with to maintain :)

**Review by Spiritaine — APPROVED** [review]
